url56b1dc76d3_wufunxiang_Libs_tgz-pJ8-org_msgpack_util_TemplatePrecompilerJ8/src/org/msgpack/template/builder/beans/StandardBeanInfo.java

```diff
@@ -0,0 +1,1505 @@
+        for (MethodDescriptor m : excludeMethods) {
+            if (method.getMethod().equals(m.getMethod())) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    @SuppressWarnings("nls")
+    private void introspectPropertyListener(Method theMethod) {
+        String methodName = theMethod.getName();
+        Class<?>[] param = theMethod.getParameterTypes();
+        if (param.length != 1) {
+            return;
+        }
+        if (methodName.equals("addPropertyChangeListener")
+                && param[0].equals(PropertyChangeListener.class))
+            canAddPropertyChangeListener = true;
+        if (methodName.equals("removePropertyChangeListener")
+                && param[0].equals(PropertyChangeListener.class))
+            canRemovePropertyChangeListener = true;
+    }
+
+    @SuppressWarnings({ "unchecked", "rawtypes" })
+    private static void introspectGet(Method theMethod,
+            HashMap<String, HashMap> propertyTable) {
+
+        String methodName = theMethod.getName();
+        int prefixLength = 0;
+        String propertyName;
+        Class propertyType;
+        Class[] paramTypes;
+        HashMap table;
+        ArrayList<Method> getters;
+
+        if (methodName == null) {
+            return;
+        }
+
+        if (methodName.startsWith(PREFIX_GET)) {
+            prefixLength = PREFIX_GET.length();
+        }
+
+        if (methodName.startsWith(PREFIX_IS)) {
+            prefixLength = PREFIX_IS.length();
+        }
+
+        if (prefixLength == 0) {
+            return;
+        }
+
+        propertyName = decapitalize(methodName.substring(prefixLength));
+
+        // validate property name
+        if (!isValidProperty(propertyName)) {
+            return;
+        }
+
+        // validate return type
+        propertyType = theMethod.getReturnType();
+
+        if (propertyType == null || propertyType == void.class) {
+            return;
+        }
+
+        // isXXX return boolean
+        if (prefixLength == 2) {
+            if (!(propertyType == boolean.class)) {
+                return;
+            }
+        }
+
+        // validate parameter types
+        paramTypes = theMethod.getParameterTypes();
+        if (paramTypes.length > 1
+                || (paramTypes.length == 1 && paramTypes[0] != int.class)) {
+            return;
+        }
+
+        table = propertyTable.get(propertyName);
+        if (table == null) {
+            table = new HashMap();
+            propertyTable.put(propertyName, table);
+        }
+
+        getters = (ArrayList<Method>) table.get(STR_GETTERS);
+        if (getters == null) {
+            getters = new ArrayList<Method>();
+            table.put(STR_GETTERS, getters);
+        }
+
+        // add current method as a valid getter
+        getters.add(theMethod);
+    }
+
+    @SuppressWarnings({ "unchecked", "rawtypes" })
+    private static void introspectSet(Method theMethod,
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 14. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test.

We recommend that you simplify this method or break it into multiple methods. For example, consider extracting the code blocks on lines 908-913, lines 915-936 into separate methods.